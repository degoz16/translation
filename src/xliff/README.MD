# xliff

rules and hooks for custom renderer to render xliff document with `translation-units` with text segments out of markdown file.

- rules: render markdown tokens as xml
- hooks: generate xml document
