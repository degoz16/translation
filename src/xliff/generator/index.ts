export {generate, TemplateParams} from './template';
export {generateTransUnit, TransUnitParams} from './trans-unit';
export {generateOpenG, generateCloseG, OpenGParams} from './g';
export {generateX} from './x';
