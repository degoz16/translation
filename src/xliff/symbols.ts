import crypto from 'crypto';

const gt = crypto.randomUUID();
const lt = crypto.randomUUID();
const sl = crypto.randomUUID();
const qt = crypto.randomUUID();

const gtre = new RegExp(`${gt}`, 'gmu');
const ltre = new RegExp(`${lt}`, 'gmu');
const qtre = new RegExp(`${qt}`, 'gmu');
const slre = new RegExp(`${sl}`, 'gmu');

function unescapeSymbols(str: string): string {
  return str.replace(gtre, '>').replace(ltre, '<').replace(qtre, '"').replace(slre, '/');
}

export {lt, gt, sl, qt, gtre, ltre, qtre, slre, unescapeSymbols};
export default {lt, gt, sl, qt, gtre, ltre, qtre, slre, unescapeSymbols};
