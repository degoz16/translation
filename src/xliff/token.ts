export type XLFToken = XLFTextToken | XLFTagToken;

export type XLFTextToken = {
    type: 'text';
    data: string;
};

export type NodeTypes = 'open' | 'close' | 'self-closing';

export type XLFTagToken = {
    type: 'tag';
    data: string;
    nodeType: NodeTypes;
    begin: string;
    end: string;
    syntax?: string;
    equivText?: string;
};

const xliffTagTokenNodeTypes = new Set(['open', 'close', 'self-closing']);

export function isXLFTextToken(token: XLFToken): token is XLFTextToken {
    return token?.type === 'text';
}

export function isXLFTagToken(token: XLFToken): token is XLFTagToken {
    return token?.type === 'tag' && xliffTagTokenNodeTypes.has(token.nodeType);
}
