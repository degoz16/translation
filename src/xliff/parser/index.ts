import assert, {ok} from 'assert';
import {XMLValidator} from 'fast-xml-parser';
import {load} from 'cheerio';
import {ChildNode, Element, isTag, isText} from 'domhandler';

import {XLFTagToken, XLFTextToken, XLFToken} from 'src/xliff/token';

export type GetTranslationsParams = {
    xliff?: string;
    units?: string[];
    useSource?: boolean;
};

const selfClosingTags = new Set(['x']);

export function parse(parameters: GetTranslationsParams): Array<Array<XLFToken>> {
    validateParams(parameters);

    const {xliff, units, useSource = false} = parameters;
    if (units) {
        return parseTargets(units.map((unit) => selectTargets(unit, useSource).get(0) as Element));
    }

    const targets = selectTargets(xliff as string, useSource);

    return parseTargets(targets.get());
}

function selectTargets(xliff: string, useSource: boolean) {
    const $ = load(xliff, {xml: true});
    const targets = $(useSource ? 'source' : 'target');

    ok(targets.length, 'Did not find any translations');

    return targets;
}

function parseTargets(targets: Element[]) {
    const parsed = new Array<Array<XLFToken>>();
    const ref = {nodes: []};

    for (const target of targets) {
        inorderNodes(target, ref);

        const tokens = nodesIntoXLFTokens(ref.nodes);
        parsed.push(tokens);

        ref.nodes = [];
    }

    return parsed;
}

function inorderNodes(node: ChildNode, ref: {nodes: ChildNode[]}) {
    if (!node) {
        return;
    }

    if (isText(node)) {
        ref.nodes.push(node);
    }

    if (isTag(node)) {
        if (selfClosingTags.has(node.name)) {
            // eslint-disable-next-line no-param-reassign
            node.attribs.nodeType = 'self-closing';
        } else {
            // eslint-disable-next-line no-param-reassign
            node.attribs.nodeType = 'open';
        }
        ref.nodes.push(node);

        let next = node.firstChild;

        while (next) {
            inorderNodes(next, ref);
            next = next.nextSibling;
        }
    }

    if (isTag(node) && !selfClosingTags.has(node.name)) {
        const closeNode = node.cloneNode();
        closeNode.attribs.nodeType = 'close';
        // @ts-ignore
        closeNode.attribs.g = node;
        ref.nodes.push(closeNode);
    }
}

function nodesIntoXLFTokens(nodes: ChildNode[]): XLFToken[] {
    const tokens = new Array<XLFToken>();

    for (const node of nodes) {
        if (isTag(node)) {
            const nodeType = node?.attribs?.nodeType;
            assert(nodeType === 'open' || nodeType === 'close' || nodeType === 'self-closing');

            const token: XLFTagToken = {
                type: 'tag',
                data: node.name,
                nodeType,
                begin: '',
                end: '',
            };

            const syntax = node?.attribs?.ctype;
            if (syntax?.length) {
                token.syntax = syntax;
            }

            const equivText = node?.attribs['equiv-text'];
            if (equivText?.length) {
                token.equivText = equivText;
            }

            // @ts-ignore
            token.g = node.attribs.g;
            token.begin = node?.attribs['x-begin'];
            token.end = node?.attribs['x-end'];

            tokens.push(token);
        } else if (isText(node)) {
            const token: XLFTextToken = {
                type: 'text',
                data: node.data,
            };
            tokens.push(token);
        }
    }

    return tokens;
}

function validateParams(parameters: GetTranslationsParams) {
    if (parameters.units) {
        parameters.units.forEach((unit) => {
            const validation = XMLValidator.validate(unit);

            if (validation !== true) {
                console.log('PROBLEM', unit)
                throw validation.err.msg;
            }
        })
        return;
    }

    const validation = XMLValidator.validate(parameters.xliff as string);

    if (validation !== true) {
        throw validation.err.msg;
    }
}
