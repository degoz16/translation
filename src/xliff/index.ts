import {render} from './renderer/md-xliff';
import {parse} from './parser';
import {generate, generateTransUnit} from './generator';

export type {XLFRenderParams, XLFRenderState} from './renderer/md-xliff/';
export type {TemplateParams, TransUnitParams} from './generator';

export const XLF = {render, parse, generate, generateTransUnit};
