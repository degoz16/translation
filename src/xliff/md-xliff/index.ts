import MarkdownIt from 'markdown-it';

import {customRenderer} from 'src/renderer';
import {token} from 'src/utils';
import {transunit} from 'src/xliff/generator';

import {hooks} from './hooks';
import {rules} from './rules';

export function render(tokens: Token[], unitId: number) {
    const xliffRenderer = new MarkdownIt({html: true});

    xliffRenderer.use(customRenderer, {rules, hooks});

    const source = xliffRenderer.renderer.render(
        [
            token('inline', {
                children: groupUselessTokens(tokens),
            }),
        ],
        xliffRenderer.options,
        {},
    );

    return transunit({
        source,
        id: unitId,
    });
}

function groupUselessTokens(tokens: Token[]) {
    const map: Record<string, Token[]> = {};
    const result = [];

    for (const part of tokens) {
        if (!part.content) {
            const [name, type] = part.type.split('_');
            if (type === 'open') {
                map[part.type] = map[part.type] || [];
                map[part.type].push(part);
            } else if (type === 'close') {
                map[name + '_open'] = map[name + '_open'] || [];
                const opener = map[name + '_open'].pop();

                if (opener) {
                    opener.g = part;
                    part.g = opener;
                }
            }
        }

        result.push(part);
    }

    return result;
}
