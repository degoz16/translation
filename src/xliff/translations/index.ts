import {ok} from 'assert';
import {XLF} from 'src/xliff';
import {XLFMDRenderer} from 'src/xliff/renderer/xliff-md';

export type TranslationsParams = {
    xliff?: string;
    units?: string[];
    useSource?: boolean;
};

export function getTranslations(parameters: TranslationsParams) {
    validateTranslationsParams(parameters);

    const translationsTokens = XLF.parse(parameters);
    const renderer = new XLFMDRenderer();

    const translations = new Map<string, string>();

    for (let i = 0; i < translationsTokens.length; i++) {
        const tokens = translationsTokens[i];

        translations.set(String(i + 1), renderer.render(tokens));
    }

    return translations;
}

function validateTranslationsParams(parameters: TranslationsParams) {
    ok(
        parameters.useSource === undefined || typeof parameters.useSource === 'boolean',
        'Unexpected useSource value',
    );
    ok(parameters.xliff || parameters.units?.length, 'Source is empty');
}
