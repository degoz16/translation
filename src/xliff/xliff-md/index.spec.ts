import {parse} from 'src/xliff';
import {extract} from 'src/api';
import {XLFToken} from 'src/xliff/token';

import {render as _render} from './index';

const tokenize = (markdown: string) => {
    const {xliff} = extract(markdown, {
        compact: true,
        source: {
            language: 'ru',
            locale: 'RU',
        },
        target: {
            language: 'en',
            locale: 'GB',
        },
    });

    const tokens = parse(xliff, {useSource: true});

    return tokens;
};

const render = (tokens: XLFToken[][]) => {
    const results = tokens.map(_render);

    return results.length === 1 ? results[0] : results;
};

describe('renders xliff to markdown', () => {
    it('renders plain text', () => {
        const tokens = tokenize('Sentence about something. Another sentence.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders strong wrapped in <x> tags', () => {
        const tokens = tokenize('Предложение номер **один**. Предложение номер **два**.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders em wrapped in <x> tags', () => {
        const tokens = tokenize('Предложение номер *один*. Предложение номер *два*.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders s wrapped in <x> tags', () => {
        const tokens = tokenize('Предложение номер ~~один~~. Предложение номер ~~два~~.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders sup wrapped in <x> tags', () => {
        const tokens = tokenize('Предложение номер ^один^. Предложение номер ^два^.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders samp wrapped in <x> tags', () => {
        const tokens = tokenize('Предложение номер ##один##. Предложение номер ##два##.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders code wrapped in <x> tag', () => {
        const tokens = tokenize('Предложение номер `один`. Предложение номер `два`.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders link with title wrapped in <x> tags', () => {
        const tokens = tokenize(
            'Предложение номер [один](one.md "one"). Предложение номер [два](two.md).',
        );

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders ref link wrapped in <x> tags', () => {
        const tokens = tokenize(
            'Предложение номер [[{#T}]](one.md "one"). Предложение номер [[{#T}]](two.md).',
        );

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders autolink wrapped in <x> tags', () => {
        const tokens = tokenize('Предложение номер один <https://www.google.com>.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders variable href link wrapped in <x> tags', () => {
        const tokens = tokenize('Предложение номер [один]({{one}} "one").');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders link containing multiple sentences wrapped in <x> tags', () => {
        const tokens = tokenize(
            '[Link text sentence one! Link text sentence Two?](file.md "Link title sentence one. Link title sentence two!").',
        );

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders image with all attributes wrapped in <x> tags', () => {
        const tokens = tokenize('Sentence ![image](image.png "hint" =100x100).');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders video wrapped in <x> tags', () => {
        const tokens = tokenize('Another sentence @[youtube](rJz4OaURJ6U)');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders heading with anchor wrapped in <x> tags', () => {
        const tokens = tokenize('# Heading with anchors {#anchor1} {#anchor2}');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders file wrapped in <x> tags', () => {
        const tokens = tokenize(
            'Sentence with {% file src="path/to/file" name="readme.md" referrerpolicy="no-referrer" rel="noopener" target="_blank" type="text/plain" %}.',
        );

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders liquid conditions wrapped in <x> tags', () => {
        const tokens = tokenize(
            'Sentence with {% if var == "val" %} val {% else %} other val {% endif %}.',
        );

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders liquid loop wrapped in <x> tags', () => {
        const tokens = tokenize('Sentence {% for x in xs %} x {% endfor %}.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders liquid function wrapped in <x> tags', () => {
        const tokens = tokenize('Sentence with function {{ user.name.slice(1, 2) }}.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders liquid filter wrapped in <x> tags', () => {
        const tokens = tokenize('Sentence with filter {{ users | length }}.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders liquid variables wrapped in <x> tags', () => {
        const tokens = tokenize('Sentence with {{ variable }}.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('renders inline html wrapped in <x> tags', () => {
        const tokens = tokenize('Sentence<br>with <b>html</b>.');

        expect(render(tokens)).toMatchSnapshot();
    });

    it('parses inline code with liquid syntax wrapped in <x> tags', () => {
        const tokens = tokenize(
            'Sentence `{{ ui-key.yacloud.common.label_tcp }} other`. Another sentence `{{user.name | capitalize}} other`.',
        );

        expect(render(tokens)).toMatchSnapshot();
    });
});
