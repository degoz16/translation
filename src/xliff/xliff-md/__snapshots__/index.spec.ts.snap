// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders xliff to markdown parses inline code with liquid syntax wrapped in <x> tags 1`] = `
Array [
  "Sentence \`{{ ui-key.yacloud.common.label_tcp }} other\`.",
  "Another sentence \`{{user.name | capitalize}} other\`.",
]
`;

exports[`renders xliff to markdown renders autolink wrapped in <x> tags 1`] =
  `"Предложение номер один <https://www.google.com>."`;

exports[`renders xliff to markdown renders code wrapped in <x> tag 1`] = `
Array [
  "Предложение номер \`один\`.",
  "Предложение номер \`два\`.",
]
`;

exports[`renders xliff to markdown renders em wrapped in <x> tags 1`] = `
Array [
  "Предложение номер *один*.",
  "Предложение номер *два*.",
]
`;

exports[`renders xliff to markdown renders file wrapped in <x> tags 1`] =
  `"Sentence with {% src=\\"path/to/file\\" name=\\"readme.md\\" referrerpolicy=\\"no-referrer\\" rel=\\"noopener\\" target=\\"_blank\\" type=\\"text/plain\\" %}."`;

exports[`renders xliff to markdown renders heading with anchor wrapped in <x> tags 1`] =
  `"Heading with anchors"`;

exports[`renders xliff to markdown renders image with all attributes wrapped in <x> tags 1`] = `
Array [
  "hint",
  "Sentence ![image](image.png \\"%%%0%%%\\" =100x100).",
]
`;

exports[`renders xliff to markdown renders inline html wrapped in <x> tags 1`] =
  `"Sentence<br>with <b>html</b>."`;

exports[
  `renders xliff to markdown renders link containing multiple sentences wrapped in <x> tags 1`
] = `
Array [
  "Link title sentence one.",
  "Link title sentence two!",
  "Link text sentence one!",
  "Link text sentence Two?](file.md \\"%%%0%%% %%%1%%%\\").",
]
`;

exports[`renders xliff to markdown renders link with title wrapped in <x> tags 1`] = `
Array [
  "one",
  "Предложение номер [один](one.md \\"%%%0%%%\\").",
  "Предложение номер [два](two.md).",
]
`;

exports[`renders xliff to markdown renders liquid conditions wrapped in <x> tags 1`] =
  `"Sentence with {% if var == \\"val\\" %} val {% else %} other val {% endif %}."`;

exports[`renders xliff to markdown renders liquid filter wrapped in <x> tags 1`] =
  `"Sentence with filter {{ users | length }}."`;

exports[`renders xliff to markdown renders liquid function wrapped in <x> tags 1`] =
  `"Sentence with function {{ user.name.slice(1, 2) }}."`;

exports[`renders xliff to markdown renders liquid loop wrapped in <x> tags 1`] =
  `"Sentence {% for x in xs %} x {% endfor %}."`;

exports[`renders xliff to markdown renders liquid variables wrapped in <x> tags 1`] =
  `"Sentence with {{ variable }}."`;

exports[`renders xliff to markdown renders plain text 1`] = `
Array [
  "Sentence about something.",
  "Another sentence.",
]
`;

exports[`renders xliff to markdown renders ref link wrapped in <x> tags 1`] = `
Array [
  "one",
  "Предложение номер [[{#T}]](one.md \\"%%%0%%%\\").",
  "Предложение номер [[{#T}]](two.md).",
]
`;

exports[`renders xliff to markdown renders s wrapped in <x> tags 1`] = `
Array [
  "Предложение номер ~~один~~.",
  "Предложение номер ~~два~~.",
]
`;

exports[`renders xliff to markdown renders samp wrapped in <x> tags 1`] = `
Array [
  "Предложение номер ##один##.",
  "Предложение номер ##два##.",
]
`;

exports[`renders xliff to markdown renders strong wrapped in <x> tags 1`] = `
Array [
  "Предложение номер **один**.",
  "Предложение номер **два**.",
]
`;

exports[`renders xliff to markdown renders sup wrapped in <x> tags 1`] = `
Array [
  "Предложение номер ^один^.",
  "Предложение номер ^два^.",
]
`;

exports[`renders xliff to markdown renders variable href link wrapped in <x> tags 1`] = `
Array [
  "one",
  "Предложение номер [один]({{one}} \\"%%%0%%%\\").",
]
`;

exports[`renders xliff to markdown renders video wrapped in <x> tags 1`] = `"Another sentence"`;
